### PR TITLE
refactor(goal_planner): use the PullOverPath, PullOverPathCandidates copied from ThreadData to reduce access

### PR DESCRIPTION
**Pull request by soblin**
## Description

- copy the pull_over_path / pull_over_path_candidates in ThreadSafeData in PullOverContextData once, in updateData to pull_over_context_data_
- all the other methods use the pull_over_path / pull_over_path_candidates in PullOverContextData instead of accessing thread_safe_data_ every time

https://tier4.atlassian.net/wiki/spaces/~621c200f9c3cce006949b60e/pages/3286696258/GoalPlanner#%E8%89%B2%E3%82%93%E3%81%AA%E7%AE%87%E6%89%80%E3%81%A7.set_foo(%E2%80%A6)%E3%82%92%E5%91%BC%E3%81%B0%E3%81%AA%E3%81%84%E3%82%88%E3%81%86%E3%81%AB%E3%81%99%E3%82%8B%EF%BC%8E

## Related links

**Parent Issue:**

- https://tier4.atlassian.net/browse/RT1-7966

## How was this PR tested?

- https://evaluation.tier4.jp/evaluation/reports/14a26394-1a2f-58dd-8741-c7b453f66576?project_id=prd_jt&state=failed
- plus https://evaluation.tier4.jp/evaluation/reports/0cd2e18f-29b5-5a17-8ffb-097ffc1f030d?project_id=prd_jt

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 1]
```diff
@@ -540,21 +540,24 @@ void GoalPlannerModule::updateData()
   path_reference_ = std::make_shared<PathWithLaneId>(getPreviousModuleOutput().reference_path);
 
   const bool found_pull_over_path = thread_safe_data_.foundPullOverPath();
-  const std::optional<PullOverPath> pull_over_path =
+  std::optional<PullOverPath> pull_over_path_recv =
```
Comment (soblin) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:543]: ここでthread_safe_data.get_pull_over_pathをコピーしてます

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 2]
```diff
@@ -540,21 +540,24 @@ void GoalPlannerModule::updateData()
...
-    *parameters_, goal_searcher_, isActivated(), pull_over_path, debug_data_.ego_polygons_expanded);
+    *parameters_, goal_searcher_, isActivated(), pull_over_path_recv,
+    debug_data_.ego_polygons_expanded);
 
   context_data_.emplace(
     path_decision_controller_.get_current_state().is_stable_safe, static_target_objects,
-    dynamic_target_objects);
+    dynamic_target_objects, std::move(pull_over_path_recv),
+    thread_safe_data_.get_pull_over_path_candidates());
+  const auto & ctx_data = context_data_.value();
```
Comment (soblin) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:560]: pull_over_path_candidates()もコピーしてきます．以降はctx_dataを使いまわします．
ほとんどの変更はthread_safe_dataへのアクセスがctx_dataへのアクセスに変わっているだけです

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 3]
```diff
@@ -1354,19 +1367,34 @@ BehaviorModuleOutput GoalPlannerModule::planPullOverAsOutput(
...
     // update thread_safe_data_
     if (path_and_goal_opt) {
       auto [pull_over_path, modified_goal] = *path_and_goal_opt;
-      deceleratePath(pull_over_path);
+      /** TODO(soblin): since thread_safe_data::pull_over_path was used as a global variable, old
+       * code was setting deceleration to thread_safe_data::pull_over_path and setOutput() accessed
+       * to the velocity profile in thread_safe_data::pull_over_path, which is a very bad usage of
+       * member variable
+       *
+       * set this selected pull_over_path to ThreadSafeData, but actually RoadParking thread does
```
Comment (kosuke55) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:1385]: minor question: road parking is equal to lane parking?

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 4]
```diff
@@ -1354,19 +1367,34 @@ BehaviorModuleOutput GoalPlannerModule::planPullOverAsOutput(
...
+       * FreespaceThreadSafeData.
+       */
       thread_safe_data_.set(goal_candidates, pull_over_path, modified_goal);
+      if (pull_over_path_with_velocity_opt) {
+        auto & pull_over_path_with_velocity = pull_over_path_with_velocity_opt.value();
+        // copy the path for later setOutput()
+        pull_over_path_with_velocity = pull_over_path;
+        // modify the velocity for latest setOutput()
+        deceleratePath(pull_over_path_with_velocity);
+      }
```
Comment (kosuke55) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:1397]: :+1:

**Review by kosuke55 — APPROVED** [review]
LGTM! thanks!

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp (hunk 3 above)
Comment (soblin) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:1385]: Yes

**Pull request merged**